### PR TITLE
Fix/handle empty model values

### PR DESCRIPTION
**Pull request by tlouisse**
fixes https://github.com/ing-bank/lion/issues/131

### CODE REVIEW

File: packages/field/src/FormatMixin.js [hunk 1]
```diff
@@ -171,9 +171,26 @@ export const FormatMixin = dedupeMixin(
...
+
+        if (typeof value !== 'string') {
+          // This means there is nothing to find inside the view that can be of
+          // interest to the Application Developer or needed to store for future form state
+          // retrieval.
+          return undefined;
+        }
+
+        const result = this.parser(value, this.formatOptions);
+        if (!result) {
```
Comment (jorenbroekema) [inline packages/field/src/FormatMixin.js]: I feel like we're doing double work here. We return a new Unparseable if the result is falsy, and then we have another check that returns a new Unparseable of result is if type 'undefined', which would already be caught by the falsy check above. We should combine the conditions into 1 conditional.

File: packages/field/src/FormatMixin.js [hunk 2]
```diff
@@ -171,9 +171,26 @@ export const FormatMixin = dedupeMixin(
       }
 
       __callParser(value = this.formattedValue) {
-        let result;
-        if (typeof value === 'string') {
-          result = this.parser(value, this.formatOptions);
+        if (value === '') {
+          // Ideally, modelValue should be undefined for empty strings.
+          // For backwards compatibility we return an empty string.
```
Comment (jorenbroekema) [inline packages/field/src/FormatMixin.js]: Can you explain the backwards compatibility? We did not have a value = empty string check before right?
I also would expect the parser to parse `''` --> `''` , is this not the case?

File: packages/field/src/FormatMixin.js [hunk 3]
```diff
@@ -171,9 +171,26 @@ export const FormatMixin = dedupeMixin(
...
+          return '';
+        }
+
+        if (typeof value !== 'string') {
```
Comment (jorenbroekema) [inline packages/field/src/FormatMixin.js]: Scenario? When could this happen? I thought native inputs values were always of type String.

File: packages/field/src/FormatMixin.js (hunk 2 above)
Comment (tlouisse) [reply packages/field/src/FormatMixin.js]: No the parser (let's say we have a date) either returns a valid Date or undefined.
Before we introduced Unparseable, it used to return '' indeed. This bugfix restores that behaviors
From a 'conceptual' perspective (whether we have a value in our model) undefined would fit better. Think we can do this small change in the next breaking release)

File: packages/field/src/FormatMixin.js (hunk 3 above)
Comment (tlouisse) [reply packages/field/src/FormatMixin.js]: More exceptional cases (extending layers) that call `__callParser` externally. Although chances are slim and it's mainly there for backwards compatibility.

File: packages/field/src/FormatMixin.js (hunk 1 above)
Comment (tlouisse) [reply packages/field/src/FormatMixin.js]: That's what happens when you have to cherry-pick commits you worked on two weeks ago.
In this branch everything was still right: https://github.com/ing-bank/lion/blob/fix/formIssues/packages/field/src/FormatMixin.js#L219

Good catch though

File: packages/field/src/FormatMixin.js (hunk 3 above)
Comment (tlouisse) [reply packages/field/src/FormatMixin.js]: In the next release it should become 
`if (typeof value !== 'string' || value = '') {
  return undefined;
}`

And I agree that the typeof check doesn't make sense if you expect a string always.
I think it also covers for edge cases (this.value is delegated to this.inputElement.value but inputElement not available yet). Defensive code that's there for backw. compat.

**Pull request merged**